interop-cats/shared/src/test/scala/zio/interop/catzSpecBase.scala

```diff
@@ -23,59 +26,123 @@ private[zio] trait catzSpecBase
     with TestInstances
     with catzSpecBaseLowPriority {
 
-  type Env = Clock with Console with System with Random
+  def checkAllAsync(name: String, f: Ticker => Laws#RuleSet): Unit =
+    checkAll(name, f(Ticker()))
+
+  def environment(implicit ticker: Ticker): ZEnv = {
+    val testClock = new Clock.Service {
+      def currentTime(unit: TimeUnit): UIO[Long] =
+        ZIO.effectTotal(unit.convert(Duration.fromScala(ticker.ctx.now())))
+
+      val currentDateTime: IO[DateTimeException, OffsetDateTime] =
+        currentTime(MILLISECONDS).map { millis =>
+          OffsetDateTime.ofInstant(Instant.ofEpochMilli(millis), ZoneOffset.UTC)
+        }
+
+      val nanoTime: UIO[Long] =
+        ZIO.effectTotal(ticker.ctx.now().toNanos)
+
+      def sleep(duration: Duration): UIO[Unit] = duration.asScala match {
+        case finiteDuration: FiniteDuration =>
+          ZIO.effectAsyncInterrupt { cb =>
+            val canceler = ticker.ctx.schedule(finiteDuration, () => cb(UIO.unit))
+            Left(UIO.effectTotal(canceler()))
+          }
+        case infinite: Infinite =>
+          ZIO.dieMessage(s"Unexpected infinite duration $infinite passed to Ticker")
+      }
+    }
+
+    ZEnv.Services.live ++ Has(testClock)
+  }
 
-  implicit def rts(implicit tc: TestContext): Runtime[Unit] = Runtime(
+  implicit def runtime(implicit ticker: Ticker): Runtime[Any] = Runtime(
     (),
     Platform
-      .fromExecutor(Executor.fromExecutionContext(Int.MaxValue)(tc))
+      .fromExecutor(Executor.fromExecutionContext(1024)(ticker.ctx))
       .withTracing(Tracing.disabled)
       .withReportFailure(_ => ())
   )
 
-  implicit val zioEqCauseNothing: Eq[Cause[Nothing]] = Eq.fromUniversalEquals
+  implicit val arbitraryAny: Arbitrary[Any] =
+    Arbitrary(Gen.const(()))
 
-  implicit def zioEqIO[E: Eq, A: Eq](implicit rts: Runtime[Any], tc: TestContext): Eq[IO[E, A]] =
-    Eq.by(_.either)
+  implicit val cogenForAny: Cogen[Any] =
+    Cogen(_.hashCode.toLong)
 
-  implicit def zioEqTask[A: Eq](implicit rts: Runtime[Any], tc: TestContext): Eq[Task[A]] =
-    Eq.by(_.either)
+  implicit def arbitraryEnvironment(implicit ticker: Ticker): Arbitrary[ZEnv] =
+    Arbitrary(Gen.const(environment))
+
+  implicit val eqForNothing: Eq[Nothing] =
+    Eq.allEqual
+
+  implicit val eqForCauseOfNothing: Eq[Cause[Nothing]] =
+    Eq.fromUniversalEquals
+
+  implicit def eqForExitOfNothing[A: Eq]: Eq[Exit[Nothing, A]] = {
+    case (Exit.Success(x), Exit.Success(y)) => x eqv y
+    case (Exit.Failure(x), Exit.Failure(y)) => x eqv y
+    case _                                  => false
+  }
 
-  implicit def zioEqUIO[A: Eq](implicit rts: Runtime[Any], tc: TestContext): Eq[UIO[A]] =
-    Eq.by(uio => taskEffectInstance.toIO(uio.sandbox.either))
+  implicit def eqForUIO[A: Eq](implicit ticker: Ticker): Eq[UIO[A]] = Eq.by { uio =>
+    try {
+      var exit: Exit[Nothing, A] = null
```

Comment:
You need to return an `Exit[Nothing, Option[A]]` here and set it to `None` initially since some tests check for no value being returned after the `TestContext` completes execution and `null` is not equal to `null`. Something like:

```scala
  def unsafeRun[A](zio: UIO[A])(implicit ticker: Ticker): Exit[Nothing, Option[A]] =
    try {

      var exit: Exit[Nothing, Option[A]] = Exit.succeed(None)

      runtime(ticker).unsafeRunAsync {
        zio
          .flatMap(ZIO.succeed(_))
          .on(ticker.ctx)
      } {
        case Exit.Success(a)     => exit = Exit.Success(Some(a))
        case Exit.Failure(cause) => exit = Exit.halt(cause)
      }

      ticker.ctx.tickAll(FiniteDuration(1L, SECONDS))

      exit
    } catch {
      case t: Throwable =>
        t.printStackTrace()
        throw t
    }

  implicit def eqForUIO[A: Eq](implicit ticker: Ticker): Eq[UIO[A]] = Eq.instance { (left, right) =>
    val l = unsafeRun(left)
    val r = unsafeRun(right)
    val p = l eqv r
    if (!p) println(s"$l was not equal to $r")
    p
  }
```

Reply:
Ah that makes sense, thanks 👍